regexp_like fusing

gortiz opened the pull request:
When executing queries like:
```sql
select col1, col2 from Table where regexp_like(col1, '/r1/') or regexp_like(col1, '/r2/')
```
Pinot has to scan the referred column twice. This PR creates an optimization that tries to fuse boolean algebra and `regexp_like` predicates. Specifically, as indicated in the Javadoc:
- Queries `where regexp_like(col1, 'r1') and regexp_like(col1, '/r2/')` will be translated to `where regexp_like(col1, '(?=r1)(?=r2)')`
- Queries `where regexp_like(col1, 'r1') or regexp_like(col1, '/r2/')` will be translated to `where regexp_like(col1, '(?:r1)|(?:r2)')`
- Queries `where not regexp_like(col1, 'r1')` will be translated to `where regexp_like(col1, '(?!r1)')`

There are some tests that apply the optimization to more advanced cases.

Regex can be quite complex. By doing some analysis I'm sure that this optimization will break some regex. For example, predicates that use backrefences may change the semantics when this optimization is applied. 

To know if the optimization can be applied or not, it would be necessary to analyze the regex, which would require to parse the regex into a AST. As far as I know Pinot doesn't have that, so this optimization is disabled by default and can be enabled by activating a new query option. There is an heuristic that applies a basic analysis to try to find incompatible expressions like backreferences. Even when the query option is enabled, the regexp that are detected as not optimizable by the heuristic will not be optimized.

Future improvements may include to translate some `text_match`, where this optimization may be useful.

PD: Originally this PR was focused on optimizing `text_match` instead of `regexp_like`. Some comments may still be referred to the original version

Jackie-Jiang commented:
Wondering how much performance gain are we expecting here? `text_match` will use the Lucene text index to get the matching docs for the request. Essentially the comparison is between 2 simple requests vs 1 combined request. Some perf number would be helpful to justify the change.

richardstartin commented:
I thought this was supposed to be targeting regexp_like

gortiz commented:
> I thought this was supposed to be targeting regexp_like

My bad. I looked for regexp in [the docs](https://docs.pinot.apache.org/configuration-reference/functions) and given that only found regexpExact, I thought it was text_match the operation we want to optimize.

Anyway, the idea is the same, so most of the code can be reused

gortiz commented on pinot-core/src/main/java/org/apache/pinot/core/plan/FilterPlanNode.java:
```diff
@@ -255,6 +255,9 @@ private BaseFilterOperator constructPhysicalOperator(FilterContext filter, int n
                 throw new UnsupportedOperationException("TEXT_MATCH is not supported on native text index");
               }
+              if (textIndexReader == null) {
+                throw new UnsupportedOperationException("TEXT_MATCH is not supported when there is no text index");
```
This moves the correctness check from execution to parsing time, which should be better, but makes `ExplainPlanQueriesTest` fail.

As indicated by its name, this tests does execute `explain plan` instead of the actual query. This is why the test didn't fail before. As this PR shouldn't be related to `text_match` but `regexp_like`, I'm going to remove this change for simplicity

richardstartin commented:
I don’t think it’s possible to express backtracking behaviour in a LIKE statement, this optimisation would target long unions of LIKE statements generated by ORMs, and should be most effective in reducing the impact on a cluster when there is no text or FST index in place. I expect this would also help FST based regular expression evaluation, so getting some numbers to justify extending to FST regex evaluators would be great.

gortiz commented:
ORMs are not going to generate them, but a user may try to do something like `regexp_like(col1, '(.+)=\1')`, which will be positive in texts like `a=a` but false in `b=a`. I think this is very bizarre, but possible. 

What I'm going to do is to try to detect the usage of backreferences in the regexp. To do so, I'm going to discard any `regexp_like` whose predicate matches with `.*(?:(?:\\[0-9]+)|(?:\\k\<\w+\>)).*`. The function may have false positives (for example the regexp`\\1` will be a false positive) and false negatives (where is more difficult to find examples, maybe playing with utf8?), but it may be good enough to activate the query option by default in a future release.

richardstartin commented:
> ORMs are not going to generate them

On the contrary, this exact pattern has been observed generated by customer systems querying Pinot

gortiz commented:
I've changed the implementation and the description of the PR to be focused on `regexp_like` instead of `text_match`. 

Anyway, I would like to have some benchmarks before this feature is merged.

richardstartin commented:
I don’t believe the native FST supports backreferences, and I don’t think it would be easy to add support either. Not sure about the lucene FST index. As mentioned above, LIKE statements (e.g. LIKE ‘%foo%’) can’t express all regular expression features, which should reduce the scope of the edge cases to be considered.

This definitely needs benchmarking for at least the following scenarios;
* Single like statement
* N like statements combined with OR

for each of:
* no FST index - both raw and dictionarized
* Lucene FST index
* Native FST index

gortiz commented:
I didn't know LIKE expressions were evaluated as REGEX_LIKE. Reading the code, it seems that the transformation is made in `RequestContextUtils`, which is evaluated after the filter optimization is done. This means that (in case I don't miss something else) the optimizations included here will not affect LIKE expressions.

I can modify the optimizer to add the ability to transform LIKE expressions into REGEX_LIKE the same way `RequestContextUtils` does right now. But I'm going to focus on the benchmark that tests the scenarios @richardstartin mentioned, but explicitly using REGEXP_LIKE for now.

Jackie-Jiang commented on pinot-perf/src/main/java/org/apache/pinot/perf/BenchmarkFuseRegexp.java:
```diff
@@ -0,0 +1,343 @@
+    long total1 = getBrokerResponse(query1).getTotalDocs();
+    long total2 = getBrokerResponse(query2).getTotalDocs();
+
+    if (total1 != total2) {
```
This is not the correct check. This returns the total documents of the table, instead of the matching docs. `getNumDocsScanned` will return the documents matched

Jackie-Jiang commented:
@gortiz Thanks for the detailed analysis. This is awesome, and very interesting!

> ## We cannot use this optimization with Lucene
> The cost is too high when the expression gets complex. It even with ` java.lang.RuntimeException: org.apache.lucene.util.automaton.TooComplexToDeterminizeException: Determinizing .*(?:^.*domain0.*$)|(?:^.*domain1.*$)|(?:^.*domain2.*$)|(?:^.*domain3.*$)|(?:^.*domain4.*$)|(?:^.*domain5.*$)|(?:^.*domain6.*$)|(?:^.*domain7.*$)|(?:^.*domain8.*$)|(?:^.*domain9.*$).* would result in more than 10000 states.` when the expression have 10 branches.
> 

Curious how native FST handles it, and why is it so much faster than the Lucene FST. cc @atris 

> ## In general `COL like 'whatever'` expression is 2x-3x worse than `regexp_like(COL, 'similar expression')`
> Reading the code it looks to me that Pinot automatically translate one to the other, but it that transformation is not applied when `BaseQueriesTest.getBrokerResponse` is called (which is great, because we can test the difference).

I think the reason is that it is not apple-to-apple comparison. `col LIKE 'abc'` is actually equivalent to `regexp_like(col, '^abc$')`

> ## Pinot regex execute faster wildcards than a single values
> It is quite clear in the optimal cases:
> 
> ```
> -- this is just a `where regexp_like(DOMAIN_NAMES, 'domain\d')`
> BenchmarkFuseRegexp.optimal10                 null  avgt    5     0.149 ±    0.082  ms/op
> -- this is just a `where regexp_like(DOMAIN_NAMES, 'domain0')`
> BenchmarkFuseRegexp.optimal1Regex             null  avgt    5   243.160 ±   59.722  ms/op
> ```
> 
> At the beginning I thought it was an error in my code, but it is easy to try that using the github_events dataset.
> 
> ```
> select * from githubEvents where regexp_like(actor_url, 'victor\d') limit 1
> ```
> 
> Responds quite faster than
> 
> ```
> select * from githubEvents where regexp_like(actor_url, 'victor7') limit 1
> ```
> 
> The difference is not the 2.000x that the benchmark shows, but is consistently between 2-3x.
> 
> I'm guessing that the issue actually comes from Java regex engine, but it is just a hunch, I didn't actually test it.

Let's re-do the benchmark. I don't believe there can be 2000x difference. Also, this is very counter-intuitive, so I'd suggest let's figure out if it is actually from the java regex engine, or because of our code.

> ### When there are several regexp in the actual expression, indexes are very useful.
> Examples:
> 
> ```
> BenchmarkFuseRegexp.decreasing9Regex        LUCENE  avgt    5     0.593 ±    0.362  ms/op
> BenchmarkFuseRegexp.decreasing9Regex        NATIVE  avgt    5     0.422 ±    0.222  ms/op
> BenchmarkFuseRegexp.decreasing9Regex          null  avgt    5  1053.673 ±  379.729  ms/op
> 
> BenchmarkFuseRegexp.increasing10Regex       LUCENE  avgt    5     0.590 ±    0.208  ms/op
> BenchmarkFuseRegexp.increasing10Regex       NATIVE  avgt    5     0.435 ±    0.159  ms/op
> BenchmarkFuseRegexp.increasing10Regex         null  avgt    5  1096.238 ±  289.928  ms/op
> ```

I feel the inefficiency is from `regexp_like(DOMAIN_NAMES, 'domain0')` (exact match) which takes much longer than expected. We should check what is causing the huge performance difference

atris commented:
Yeah, the REGEXP_LIKE translation from LIKE is not equivalent.

Also, remember that any FST traversal will benefit most from a selective expression rather than a full term match.

Lastly, why do you say that indices don't help for single regexes? Unless I am misreading the benchmark result, even for a single regex case, native index is pretty fast?

atris commented:
I will check out the branch and give some comments tomorrow

richardstartin commented:
> Thanks @atris for taking a look. I also have a question around how Lucene and native FST handle the query, specifically for these results:
> 
> ```
> SELECT INT_COL FROM MyTable WHERE regexp_like(DOMAIN_NAMES, '(?:^.*domain9.*$)|...|(?:^.*domain1.*$)')
> BenchmarkFuseRegexp.decreasing9Fusing       LUCENE  avgt    5    24.198 ±   13.439  ms/op
> BenchmarkFuseRegexp.decreasing9Fusing       NATIVE  avgt    5     0.263 ±    0.019  ms/op
> BenchmarkFuseRegexp.decreasing9Fusing         null  avgt    5     0.476 ±    0.040  ms/op
> 
> SELECT INT_COL FROM MyTable WHERE regexp_like(DOMAIN_NAMES, '(?:^.*domain0.*$)|...|(?:^.*domain9.*$)')
> BenchmarkFuseRegexp.increasing10Fusing      LUCENE  avgt    5     ERROR
> BenchmarkFuseRegexp.increasing10Fusing      NATIVE  avgt    5     0.326 ±    0.138  ms/op
> BenchmarkFuseRegexp.increasing10Fusing        null  avgt    5     0.226 ±    0.062  ms/op
> ```
> 
> We can see that Lucene FST is very slow in this case, but native is still performing okay. Trying to understand what is causing this difference.

The difference will be due to determinization of the automaton, which the native implementation doesn’t do.

atris commented:
@Jackie-Jiang one thing is that the native index aggressively tries to prune branches, both while building the query automaton and during the traversal.

 That is also what causes general suffix and prefix queries to be much faster.

richardstartin commented:
> > > Thanks @atris for taking a look. I also have a question around how Lucene and native FST handle the query, specifically for these results:
> > > ```
> > > SELECT INT_COL FROM MyTable WHERE regexp_like(DOMAIN_NAMES, '(?:^.*domain9.*$)|...|(?:^.*domain1.*$)')
> > > BenchmarkFuseRegexp.decreasing9Fusing       LUCENE  avgt    5    24.198 ±   13.439  ms/op
> > > BenchmarkFuseRegexp.decreasing9Fusing       NATIVE  avgt    5     0.263 ±    0.019  ms/op
> > > BenchmarkFuseRegexp.decreasing9Fusing         null  avgt    5     0.476 ±    0.040  ms/op
> > > 
> > > SELECT INT_COL FROM MyTable WHERE regexp_like(DOMAIN_NAMES, '(?:^.*domain0.*$)|...|(?:^.*domain9.*$)')
> > > BenchmarkFuseRegexp.increasing10Fusing      LUCENE  avgt    5     ERROR
> > > BenchmarkFuseRegexp.increasing10Fusing      NATIVE  avgt    5     0.326 ±    0.138  ms/op
> > > BenchmarkFuseRegexp.increasing10Fusing        null  avgt    5     0.226 ±    0.062  ms/op
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > We can see that Lucene FST is very slow in this case, but native is still performing okay. Trying to understand what is causing this difference.
> > 
> > 
> > The difference will be due to determinization of the automaton, which the native implementation doesn’t do.
> 
> I doubt determinization alone can cause such a massive difference.

Well, it’s an exponential time algorithm and Lucene refuses to even attempt to determinize some of the automata in the benchmark because they are too large. For an apples to apples comparison, native determinization should be re-enabled (by temporarily reverting/reversing #8237)

atris commented:
> > > > Thanks @atris for taking a look. I also have a question around how Lucene and native FST handle the query, specifically for these results:
> > > > ```
> > > > SELECT INT_COL FROM MyTable WHERE regexp_like(DOMAIN_NAMES, '(?:^.*domain9.*$)|...|(?:^.*domain1.*$)')
> > > > BenchmarkFuseRegexp.decreasing9Fusing       LUCENE  avgt    5    24.198 ±   13.439  ms/op
> > > > BenchmarkFuseRegexp.decreasing9Fusing       NATIVE  avgt    5     0.263 ±    0.019  ms/op
> > > > BenchmarkFuseRegexp.decreasing9Fusing         null  avgt    5     0.476 ±    0.040  ms/op
> > > > 
> > > > SELECT INT_COL FROM MyTable WHERE regexp_like(DOMAIN_NAMES, '(?:^.*domain0.*$)|...|(?:^.*domain9.*$)')
> > > > BenchmarkFuseRegexp.increasing10Fusing      LUCENE  avgt    5     ERROR
> > > > BenchmarkFuseRegexp.increasing10Fusing      NATIVE  avgt    5     0.326 ±    0.138  ms/op
> > > > BenchmarkFuseRegexp.increasing10Fusing        null  avgt    5     0.226 ±    0.062  ms/op
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > We can see that Lucene FST is very slow in this case, but native is still performing okay. Trying to understand what is causing this difference.
> > > 
> > > 
> > > The difference will be due to determinization of the automaton, which the native implementation doesn’t do.
> > 
> > I doubt determinization alone can cause such a massive difference.
> 
> Well, it’s an exponential time algorithm and Lucene refuses to even attempt to determinize some of the automata in the benchmark because they are too large. For an apples to apples comparison, native determinization should be re-enabled (by temporarily reverting/reversing #8237)

I am still failing to understand what the concern is. 

Even if its plain determinization that causes native to be significantly faster, is that not a good thing?

richardstartin commented:
> > > > > Thanks @atris for taking a look. I also have a question around how Lucene and native FST handle the query, specifically for these results:
> > > > > ```
> > > > > SELECT INT_COL FROM MyTable WHERE regexp_like(DOMAIN_NAMES, '(?:^.*domain9.*$)|...|(?:^.*domain1.*$)')
> > > > > BenchmarkFuseRegexp.decreasing9Fusing       LUCENE  avgt    5    24.198 ±   13.439  ms/op
> > > > > BenchmarkFuseRegexp.decreasing9Fusing       NATIVE  avgt    5     0.263 ±    0.019  ms/op
> > > > > BenchmarkFuseRegexp.decreasing9Fusing         null  avgt    5     0.476 ±    0.040  ms/op
> > > > > 
> > > > > SELECT INT_COL FROM MyTable WHERE regexp_like(DOMAIN_NAMES, '(?:^.*domain0.*$)|...|(?:^.*domain9.*$)')
> > > > > BenchmarkFuseRegexp.increasing10Fusing      LUCENE  avgt    5     ERROR
> > > > > BenchmarkFuseRegexp.increasing10Fusing      NATIVE  avgt    5     0.326 ±    0.138  ms/op
> > > > > BenchmarkFuseRegexp.increasing10Fusing        null  avgt    5     0.226 ±    0.062  ms/op
> > > > > ```
> > > > > 
> > > > > 
> > > > >     
> > > > >       
> > > > >     
> > > > > 
> > > > >       
> > > > >     
> > > > > 
> > > > >     
> > > > >   
> > > > > We can see that Lucene FST is very slow in this case, but native is still performing okay. Trying to understand what is causing this difference.
> > > > 
> > > > 
> > > > The difference will be due to determinization of the automaton, which the native implementation doesn’t do.
> > > 
> > > 
> > > I doubt determinization alone can cause such a massive difference.
> > 
> > 
> > Well, it’s an exponential time algorithm and Lucene refuses to even attempt to determinize some of the automata in the benchmark because they are too large. For an apples to apples comparison, native determinization should be re-enabled (by temporarily reverting/reversing #8237)
> 
> I am still failing to understand what the concern is.
> 
> Even if its plain determinization that causes native to be significantly faster, is that not a good thing?

@Jackie-Jiang was asking about the root cause of the difference. I suggested that the root cause is likely to be determinization, but it needs to be investigated. Where has concern been expressed?

richardstartin commented:
Just circling back here having taken a quick look at `BenchmarkFuseRegexp.decreasing9Fusing` - I got similar results to Gonzalo:

```
Benchmark                              (_fstType)  Mode  Cnt   Score   Error  Units
BenchmarkFuseRegexp.decreasing9Fusing      LUCENE  avgt    5  32.525 ± 2.733  ms/op
BenchmarkFuseRegexp.decreasing9Fusing      NATIVE  avgt    5   0.487 ± 0.123  ms/op
BenchmarkFuseRegexp.decreasing9Fusing        null  avgt    5   0.845 ± 0.204  ms/op
```

 Lucene spends virtually all of its time in _minimization_ (which I mistakenly referred to as _determinization_ earlier):

<img width="936" alt="Screenshot 2022-06-13 at 21 10 51" src="https://user-images.githubusercontent.com/16439049/173436813-a13a5830-ad7b-4a8c-8e5f-3cf8003640df.png">
 
Now if I enable minimization in the native implementation by reverting #8237, the scores change drastically:

```
Benchmark                              (_fstType)  Mode  Cnt    Score    Error  Units
BenchmarkFuseRegexp.decreasing9Fusing      LUCENE  avgt    5   32.399 ±  3.005  ms/op
BenchmarkFuseRegexp.decreasing9Fusing      NATIVE  avgt    5  366.520 ± 40.574  ms/op
BenchmarkFuseRegexp.decreasing9Fusing        null  avgt    5    0.819 ±  0.172  ms/op
```

As do the profiles. Without minimization, there is no clear single bottleneck:
<img width="879" alt="Screenshot 2022-06-13 at 21 14 47" src="https://user-images.githubusercontent.com/16439049/173437464-980859ce-e7c5-4e8f-b43e-ba609ab39396.png">

With minimization, there is a single bottleneck, and it's in minimization :

<img width="931" alt="Screenshot 2022-06-13 at 21 23 10" src="https://user-images.githubusercontent.com/16439049/173438712-5f8efc36-1f8d-4ea5-8bb6-b135d21ff4fc.png">


As to whether this is cause for concern is beside the point; this is the root cause of the difference observed, rather than a quirk of the benchmark which needs to be analysed and removed.

gortiz commented:
> Lastly, why do you say that indices don't help for single regexes? Unless I am misreading the benchmark result, even for a single regex case, native index is pretty fast?

My benchmark may not be the best and it is only measuring a few regexp cases, but I wouldn't say that native gives us a significant advantage:

```
-- this actually has a single, quite complex regex expression
BenchmarkFuseRegexp.decreasing9Fusing       LUCENE  avgt    5    24.198 ±   13.439  ms/op
BenchmarkFuseRegexp.decreasing9Fusing       NATIVE  avgt    5     0.263 ±    0.019  ms/op <-- here native is faster by 1.8x
BenchmarkFuseRegexp.decreasing9Fusing         null  avgt    5     0.476 ±    0.040  ms/op
-- this actually has a single, quite complex regex expression
BenchmarkFuseRegexp.increasing10Fusing      LUCENE  avgt    5     ERROR
BenchmarkFuseRegexp.increasing10Fusing      NATIVE  avgt    5     0.326 ±    0.138  ms/op <-- here native is slower by 1.4x, although the difference is lower than the error margin
BenchmarkFuseRegexp.increasing10Fusing        null  avgt    5     0.226 ±    0.062  ms/op
-- this is just a `where regexp_like(DOMAIN_NAMES, 'domain\d')`
BenchmarkFuseRegexp.optimal10               LUCENE  avgt    5     0.169 ±    0.066  ms/op
BenchmarkFuseRegexp.optimal10               NATIVE  avgt    5     0.136 ±    0.011  ms/op  <-- here native is faster by 1.1x, although the difference is lower than the error margin
BenchmarkFuseRegexp.optimal10                 null  avgt    5     0.149 ±    0.082  ms/op
```

So we have one case where native is 1.8x faster and some others were it is equivalent to the java regex engine, which is known by its lack of performance. I mean, it is not like the native index is atrocious, but I would expect more from an index than a situational 1.8x performance increase. 

To be clear, as I already said, this is not a benchmark on the FST itself and it is very narrow, so I don't want to get conclusions about whether we should or we should not use these indexes. I think this benchmark proves that there are very specific cases where these indexes are not very useful and therefore it would be nice to invest time to design proper tests with different kind of regex to verify whether FST indexes are in general useful evaluating regex or not.

gortiz commented:
I was able to invest more time in this topic today and I have to say that my initial benchmark was completely useless. Apart from some typos in the queries (which may have some impact), there were two problems that completely fooled me:
- The implicit limit introduced by Pinot. As only the first 10 elements were returned, queries that actually found elements could finish faster.
- Something is wrong with Lucene and Native indexes. I guess I'm not configuring them correctly, but they simply return 0 results when the regexp is a little bit complex (like having groups).

Therefore I've changed all the tests to return a count and each iteration is actually verifying whether the expected value is returned or not, in which case the benchmark is stopped. As a result, all fusing tests fail whenever there is an index. I hope it is something related to my test (ie I didn't configure the index properly) and not some semantic discrepancy these indexes have in relation how each engine evaluates regexp, as that could mean that there are customers receiving false results.

These are the results I've got. Note that some combinations are not shown. For example, decreasing9Fusing and LUCENE or NATIVE. That means that these combinations did not return what it was expected (in all these particular cases it means that they return 0 rows). Note that apart from the fusing case, benchmarks like `optimal10`, which is `regexp_like(DOMAIN_NAMES, 'domain\d')` did also failed.

```
Benchmark                                  (_fstType)  Mode  Cnt     Score      Error  Units
BenchmarkFuseRegexp.decreasing9Fusing            null  avgt    5  1438.185 ±   73.963  ms/op
BenchmarkFuseRegexp.decreasing9Like            LUCENE  avgt    5    41.927 ±    0.984  ms/op
BenchmarkFuseRegexp.decreasing9Like            NATIVE  avgt    5    40.232 ±    5.597  ms/op
BenchmarkFuseRegexp.decreasing9Like              null  avgt    5  3957.902 ±  154.228  ms/op
BenchmarkFuseRegexp.decreasing9Regex           LUCENE  avgt    5    40.744 ±    2.059  ms/op
BenchmarkFuseRegexp.decreasing9Regex           NATIVE  avgt    5    61.443 ±    1.301  ms/op
BenchmarkFuseRegexp.decreasing9Regex             null  avgt    5  2116.747 ±  602.570  ms/op
BenchmarkFuseRegexp.increasing10Fusing           null  avgt    5  1320.869 ±   45.509  ms/op
BenchmarkFuseRegexp.increasing10Like           LUCENE  avgt    5    42.253 ±    0.695  ms/op
BenchmarkFuseRegexp.increasing10Like           NATIVE  avgt    5    39.049 ±    1.829  ms/op
BenchmarkFuseRegexp.increasing10Like             null  avgt    5  5071.207 ± 3134.887  ms/op
BenchmarkFuseRegexp.increasing10Regex          LUCENE  avgt    5    44.084 ±    2.737  ms/op
BenchmarkFuseRegexp.increasing10Regex          NATIVE  avgt    5    39.650 ±    1.567  ms/op
BenchmarkFuseRegexp.increasing10Regex            null  avgt    5  2318.479 ±   75.632  ms/op
BenchmarkFuseRegexp.optimal10                    null  avgt    5   185.757 ±    3.426  ms/op
BenchmarkFuseRegexp.optimal10NotFound          LUCENE  avgt    5     0.212 ±    0.136  ms/op
BenchmarkFuseRegexp.optimal10NotFound          NATIVE  avgt    5     0.201 ±    0.140  ms/op
BenchmarkFuseRegexp.optimal10NotFound            null  avgt    5   248.525 ±   60.819  ms/op
BenchmarkFuseRegexp.optimal1Like               LUCENE  avgt    5    16.431 ±    3.533  ms/op
BenchmarkFuseRegexp.optimal1Like               NATIVE  avgt    5    16.586 ±    0.828  ms/op
BenchmarkFuseRegexp.optimal1Like                 null  avgt    5   468.731 ±  185.623  ms/op
BenchmarkFuseRegexp.optimal1LikeNotFound       LUCENE  avgt    5     0.197 ±    0.062  ms/op
BenchmarkFuseRegexp.optimal1LikeNotFound       NATIVE  avgt    5     0.170 ±    0.072  ms/op
BenchmarkFuseRegexp.optimal1LikeNotFound         null  avgt    5   477.176 ±  129.445  ms/op
BenchmarkFuseRegexp.optimal1Regex              LUCENE  avgt    5    15.485 ±    0.517  ms/op
BenchmarkFuseRegexp.optimal1Regex              NATIVE  avgt    5    16.711 ±    0.857  ms/op
BenchmarkFuseRegexp.optimal1Regex                null  avgt    5   205.676 ±    6.921  ms/op
BenchmarkFuseRegexp.optimal1RegexNotFound      LUCENE  avgt    5     0.212 ±    0.123  ms/op
BenchmarkFuseRegexp.optimal1RegexNotFound      NATIVE  avgt    5     0.186 ±    0.029  ms/op
BenchmarkFuseRegexp.optimal1RegexNotFound        null  avgt    5   239.459 ±   13.846  ms/op
BenchmarkFuseRegexp.selective2Fusing             null  avgt    5   541.530 ±  168.173  ms/op
BenchmarkFuseRegexp.selective2Like             LUCENE  avgt    5    19.961 ±    0.937  ms/op
BenchmarkFuseRegexp.selective2Like             NATIVE  avgt    5    20.387 ±    1.953  ms/op
BenchmarkFuseRegexp.selective2Like               null  avgt    5   916.620 ±   96.702  ms/op
BenchmarkFuseRegexp.selective2Regex            LUCENE  avgt    5    23.366 ±    0.795  ms/op
BenchmarkFuseRegexp.selective2Regex            NATIVE  avgt    5    20.037 ±    0.601  ms/op
BenchmarkFuseRegexp.selective2Regex              null  avgt    5   389.965 ±    7.284  ms/op
```

I've also changed `RegexpPatternConverterUtils` (and its relative test) to do not introduce useless `^.*` at the beginning of the expression nor `.*$` at the end. I've repeated the benchmark with that change and these are the results I've got. As you can see, most benchmarks didn't change that much, but the ones that use like and have no index are twice as fast, so they achieve the same numbers than their equivalent `optimalXRegex`, as expected.

```
Benchmark                                  (_fstType)  Mode  Cnt     Score     Error  Units
BenchmarkFuseRegexp.decreasing9Fusing            null  avgt    5  1365.403 ±  91.822  ms/op
BenchmarkFuseRegexp.decreasing9Like            LUCENE  avgt    5    40.553 ±   1.394  ms/op
BenchmarkFuseRegexp.decreasing9Like            NATIVE  avgt    5    39.057 ±   0.905  ms/op
BenchmarkFuseRegexp.decreasing9Like              null  avgt    5  2046.090 ±  40.428  ms/op
BenchmarkFuseRegexp.decreasing9Regex           LUCENE  avgt    5    42.584 ±   2.387  ms/op
BenchmarkFuseRegexp.decreasing9Regex           NATIVE  avgt    5    61.583 ±   1.893  ms/op
BenchmarkFuseRegexp.decreasing9Regex             null  avgt    5  2054.523 ±  75.452  ms/op
BenchmarkFuseRegexp.increasing10Fusing           null  avgt    5  1305.249 ±  65.576  ms/op
BenchmarkFuseRegexp.increasing10Like           LUCENE  avgt    5    43.328 ±   2.040  ms/op
BenchmarkFuseRegexp.increasing10Like           NATIVE  avgt    5    61.580 ±   1.149  ms/op
BenchmarkFuseRegexp.increasing10Like             null  avgt    5  2272.112 ±  99.393  ms/op
BenchmarkFuseRegexp.increasing10Regex          LUCENE  avgt    5    54.319 ±   2.048  ms/op
BenchmarkFuseRegexp.increasing10Regex          NATIVE  avgt    5    39.080 ±   1.817  ms/op
BenchmarkFuseRegexp.increasing10Regex            null  avgt    5  2267.062 ±  49.525  ms/op
BenchmarkFuseRegexp.optimal10                    null  avgt    5   187.012 ±  13.054  ms/op
BenchmarkFuseRegexp.optimal10NotFound          LUCENE  avgt    5     0.210 ±   0.103  ms/op
BenchmarkFuseRegexp.optimal10NotFound          NATIVE  avgt    5     0.201 ±   0.066  ms/op
BenchmarkFuseRegexp.optimal10NotFound            null  avgt    5   242.423 ±  17.612  ms/op
BenchmarkFuseRegexp.optimal1Like               LUCENE  avgt    5    16.395 ±   0.884  ms/op
BenchmarkFuseRegexp.optimal1Like               NATIVE  avgt    5    15.506 ±   0.591  ms/op
BenchmarkFuseRegexp.optimal1Like                 null  avgt    5   209.389 ±  11.456  ms/op
BenchmarkFuseRegexp.optimal1LikeNotFound       LUCENE  avgt    5     0.209 ±   0.129  ms/op
BenchmarkFuseRegexp.optimal1LikeNotFound       NATIVE  avgt    5     0.172 ±   0.021  ms/op
BenchmarkFuseRegexp.optimal1LikeNotFound         null  avgt    5   245.542 ±  35.323  ms/op
BenchmarkFuseRegexp.optimal1Regex              LUCENE  avgt    5    15.421 ±   0.804  ms/op
BenchmarkFuseRegexp.optimal1Regex              NATIVE  avgt    5    15.908 ±   0.491  ms/op
BenchmarkFuseRegexp.optimal1Regex                null  avgt    5   204.505 ±  13.134  ms/op
BenchmarkFuseRegexp.optimal1RegexNotFound      LUCENE  avgt    5     0.210 ±   0.089  ms/op
BenchmarkFuseRegexp.optimal1RegexNotFound      NATIVE  avgt    5     0.203 ±   0.107  ms/op
BenchmarkFuseRegexp.optimal1RegexNotFound        null  avgt    5   235.262 ±   8.533  ms/op
BenchmarkFuseRegexp.selective2Fusing             null  avgt    5   517.545 ± 193.977  ms/op
BenchmarkFuseRegexp.selective2Like             LUCENE  avgt    5    23.622 ±   1.178  ms/op
BenchmarkFuseRegexp.selective2Like             NATIVE  avgt    5    20.547 ±   0.778  ms/op
BenchmarkFuseRegexp.selective2Like               null  avgt    5   399.638 ±  54.746  ms/op
BenchmarkFuseRegexp.selective2Regex            LUCENE  avgt    5    23.883 ±   0.802  ms/op
BenchmarkFuseRegexp.selective2Regex            NATIVE  avgt    5    20.113 ±   0.858  ms/op
BenchmarkFuseRegexp.selective2Regex              null  avgt    5   416.987 ±  28.789  ms/op
```

With this results, the optimization introduced in this PR doesn't seem to be amazing, but to be honest I'm quite worried about what is going on with regexp and Lucene/Native indexes. As said above, I hope that the benchmark is not correctly configuring the indexes, but even with that, it seems very dangerous that an index can be configured in such a way that it affects the semantics of the queries.

@Jackie-Jiang @atris I would really appreciate if any of you can take a look at the benchmark trying to find what is going on.

atris commented:
@gortiz are you creating a FST index or a text index?

atris commented:
@gortiz another thing to remember is that neither Lucene nor native FST index follow the java regex pattern language.

 If you look up the Lucene regex documentation, and look at the FST index and LIKE tests in Pinot, you will notice the difference.

Unfortunately, since we do not actually parse the regex engine until deep inside the FST territory, we have no way of reporting unknown characters. That may be a factor leading to 0 results for your queries

gortiz commented:
> @gortiz another thing to remember is that neither Lucene nor native FST index follow the java regex pattern language.
> 
> If you look up the Lucene regex documentation, and look at the FST index and LIKE tests in Pinot, you will notice the difference.
> 
> Unfortunately, since we do not actually parse the regex engine until deep inside the FST territory, we have no way of reporting unknown characters. That may be a factor leading to 0 results for your queries

Wow, that is shocking and I would see that as a huge problem to use Pinot in a big team. In cases like that I would expect to have a new operation that does the search using the Lucene expression language or whatever we want to use and fail if there is no index. What you are saying is that if I have a column that is not indexed and suddenly a colleague of mine decides to index that column, queries I had will start to return different (and incorrect) results. 

Anyway, if that is what was decided, that is what we have. I'm going to close this PR, as it doesn't make sense to apply this optimization if it can be produce false results when someone adds an index.

atris commented:
@gortiz There has been a long comment trail on this PR so I am summarising my responses in this comment:

1. Performance -- Like any other index, FST indices also have their worst case. FST is an automaton, and in an extreme scenario where a high number of branches are traversed for a query, the performance of the index will approach the brute force approximation.

In this case, the query you tried had an abnormally large number of branches, to the extent that Lucene decided to not even try and process the query. The average user rarely fires such queries, and in the corner cases where they do, they should be ok with a brute force latency, as long as their 99th percentile queries keep performing significantly better, which is the case as is evident from the benchmarks that test the FST index.

2. Language Semantics -- Lucene has, for long, published its regex dialect clearly. Engines based on Lucene, such as ElasticSearch and Apache Solr, follow the same dialect (https://help.oncrawl.com/en/articles/1685982-lucene-regex-cheat-sheet)

It is arguable whether this is correct or not, but we have to keep historical perspective in mind, since many Pinot users migrate from one of these engines.

Please note that, from native FST index's perspective, it is pretty simple to support java regex language. The reason I did not do that was because the existing FST index was based on Lucene and thus supported Lucene's regex dialect. For back compatibility, native FST index supports the same dialect.

3. Parsing And Reporting Invalid Regexes: I agree that in an ideal scenario, we should be parsing regexes and reporting characters that we do not accept. However, given the complexity of an average regex, the time complexity of doing this will be non trivial.

4. Specific To This PR: I looked into the PR and while the effort is appreciative, I do not see significant performance improvement. Also, the API surface changes required for this change are high, which makes me a bit uneasy.

gortiz commented:
Yeah, I completely understand that we should maintain the Lucene language when Natife FST was created. My concern is with the fact that when Lucene index was added, we decided to accept a different language in `regexp_match`. IMHO that is a consistency problem that is not easy to explain to Pinot customers. Pinot doesn't include documentation about this operation and how it changes depending on whether there is an index or not and I feel sorry for whoever has to write that doc, because it seems quite difficult to describe.

IMHO a better solution would be to do not optimize `regexp_match` with indexes and to have another operation whose string argument was a Lucene expression.

It is too late to do that, so let's focus on useful things. This PR is closed, but I'm going to cherry pick the like to regex optimization to add it in another PR.
